### PR TITLE
Force reset password

### PR DESCRIPTION
**Pull request by ghost**
Added a functionality that forces users to reset their password on first login, if the initial password was set by the admin

### CODE REVIEW

**Comment by gary-kim** [comment]
I think it would be nice to have some way to set a password without requiring a password reset in the interface.

File: apps/provisioning_api/lib/Controller/UsersController.php [hunk 1]
```diff
@@ -594,6 +598,9 @@ public function editUser(string $userId, string $key, string $value): DataRespon
 					$this->accountManager->updateUser($targetUser, $userAccount);
 				}
 				break;
+			case 'initial':
```
Comment (kesselb) [inline apps/provisioning_api/lib/Controller/UsersController.php]: Please create a constant like `AccountManager::PROPERTY_INITIAL`

File: apps/provisioning_api/lib/Controller/UsersController.php [hunk 2]
```diff
@@ -520,6 +523,7 @@ public function editUser(string $userId, string $key, string $value): DataRespon
 				$permittedFields[] = AccountManager::PROPERTY_WEBSITE;
 				$permittedFields[] = AccountManager::PROPERTY_TWITTER;
 				$permittedFields[] = 'quota';
+				$permittedFields[] = 'initial';
```
Comment (kesselb) [inline apps/provisioning_api/lib/Controller/UsersController.php]: Could you find something more specific? Perhaps `forcePasswordReset`? `initial` is very generic.

File: core/Controller/LoginController.php [hunk 3]
```diff
@@ -306,6 +306,13 @@ public function tryLogin(string $user,
 				$result->getErrorMessage()
 			);
 		}
+		
+		if($this->config->getUserValue($user, 'core', 'initial') === 'true') {
+			$token = $this->config->getUserKeys($user, 'login_token')[0];
```
Comment (kesselb) [inline core/Controller/LoginController.php]: OK. The right way to do this:

Take the password reset token logic from `NewUserMailHelper`. Perhaps we need to find a better way for this. A generatePasswordResetToken method. If the user login and `initial` is set generate a password reset token and forward the user.

@rullzer @ChristophWurst

File: apps/provisioning_api/lib/Controller/UsersController.php [hunk 4]
```diff
@@ -349,6 +349,9 @@ public function addUser(string $userid,
 						'app' => 'ocs_api',
 					]);
 				}
+			} else {
+				//Password was provided by the admin
```
Comment (kesselb) [inline apps/provisioning_api/lib/Controller/UsersController.php]: Case 1: No password, No email = failure
Case 2: No password, Email given = username and password reset link is mailed to user
Case 3: Password given, Email given = username is mailed to user
_Case 4: Password given, No email = password reset forced after login_

At least that must be documented somewhere. I'm against any more magic here. I don't think the behaviour is guessable or intuitive even without this pr. We should explain what Nextcloud is actually are going to do and/or give the admin a change to pick the action. 

@jancborchardt

**Review by kesselb — COMMENTED** [review]
Thank you for picking this up :+1: 

I'm sorry for all those remarks. Please wait for another review for a second opinion. I only know the authentication in general. Probably there are better ways than my suggestions (especially for the tokens).

File: apps/provisioning_api/lib/Controller/UsersController.php (hunk 4 above)
Comment (kesselb) [reply apps/provisioning_api/lib/Controller/UsersController.php]: Also Password given, No email is what I usually do for developing :rofl: Need to test something with a different user. Create user with  password and no email, Login => Forced password reset :angry: 

I know that's unusual but this PR is a behavioural change. I'm sure there is someone out there happy with the current approach. People will complain about this change if there is no way to turn it off. 

For example: A self-service for account creation. People are able to create a user account for Nextcloud with a internal tool. Password is only visible to user. With this change they have to reset their password again for a account just created seconds ago.

**Comment by skjnldsv** [comment]
As there is no feedback since a while I will close this PR. If you are still willing to get this in, please address the potential comments and rebase to latest master. Then, feel free to re-open.

**Comment by delacosta456** [comment]
hi @ein-giga-self 
while looking for something similar to what you have listed in your feature i cam to this post.
I would like to know what is the exact state of those functionalities .. Did you make another PR about these somewhere else ?
Thanks

**Pull request closed without merging**